[CDAP-17239] Fix AuthorizerInstantiator Classloading Bug

dli357 opened the pull request:
See [CDAP-17239](https://issues.cask.co/browse/CDAP-17239) and [CDAP-17240](https://issues.cask.co/browse/CDAP-17240).

In CDAP [StandaloneMain](https://github.com/cdapio/cdap/blob/d69fc4b0eb0691156e515343e0fa87e72e409471/cdap-standalone/src/main/java/io/cdap/cdap/StandaloneMain.java), the AuthorizerInstantiator unpacks the custom authorization extension JAR file into a folder in `data/tmp`. However, the AuthorizerInstantiator is actually created in the constructor for StandaloneMain, and then cleanupTempDir() is called in startup(). This causes the unpacked JAR files to be deleted immediately after instantiation.

ClassLoaders generally perform lazy loading, which means that if an authorization extension attempts to load a class after this runtime classpath is deleted, it will throw a ClassNotFoundException.

shifuxu0301 reviewed (commented):
Let's add more details in the commit description to understand why this order change will help fix the issue.

shifuxu0301 commented on cdap-standalone/src/main/java/io/cdap/cdap/StandaloneMain.java:
```diff
@@ -162,6 +162,8 @@ private StandaloneMain(List<Module> modules, CConfiguration cConf) {
     Thread.setDefaultUncaughtExceptionHandler(new UncaughtExceptionHandler());
     this.cConf = cConf;
 
+    cleanupTempDir();
```
Can we put a comment here to explain what we expect for this cleanup in constructor.

dli357 replied:
Fixed

albertshau replied:
normally it's not a good idea to perform work in the constructor. I'm also confused how this fixes the problem, is the classloader mentioned in the bug created before Standalone is started?

dli357 replied:
I agree with moving the logic out of the constructor. Currently, StandaloneMain's constructor creates several other classes which eventually instantiates an instance of Authorizer, causing the AuthorizerInstantiator to unpack the provided JAR file into the tmp directory. Since cleanupTempDir is currently called in startup, it is called after the constructor finishes, deleting the loaded classpath.

albertshau commented:
It's usually not a good idea to perform work or do something that has side effects in a constructor. It seems like it would be more appropriate for more logic to be moved to startup time rather than moving cleanup from start up to the constructor.

dli357 committed:
[CDAP-17239] Fix classloading bug where the classpath directory is deleted in StandaloneMain

    [CDAP-17239] Added comment

    [CDAP-17239] Moved temporary directory cleanup logic out of constructor

[CDAP-17239] Construct tmpDir outside of static method

wyzhang commented on cdap-standalone/src/main/java/io/cdap/cdap/StandaloneMain.java:
```diff
@@ -392,7 +390,7 @@ public void shutDown() {
+  private static void cleanupTempDir(CConfiguration cConf) {
     File tmpDir = new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
                            cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
 
```
Construct the path at callsite and pass that in, so it is clear what state the function needs instead of passing the whole configuration structure that this function doesn't care about the other n-2 fields

dli357 replied:
Fixed
